esp-hal/src/dma/buffers.rs

```diff
@@ -2,28 +2,95 @@ use core::ptr::null_mut;
 
 use super::*;
 use crate::soc::is_slice_in_dram;
-#[cfg(esp32s3)]
-use crate::soc::is_slice_in_psram;
+#[cfg(psram_dma)]
+use crate::soc::{is_slice_in_psram, is_valid_psram_address};
+
+/// Burst transfer configuration.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum BurstTransfer {
```

Comment:
So I sketched this up, but I immediately want to do a revision based on some of what I've learnt:

S2/S3 external DMA is an interesting mix of P4's DMAs: for internal transfers, they don't force burst mode (like AHB-DMA), for external transfers they do (like AXI DMA).

I'm considering turning this enum into a 2-field struct, that has different field names on the different MCUs:

S2/S3: internal_burst and external_burst
P4: ahb_burst and axi_burst

I would repurpose the current enum as InternalBurstTransfer or AhbBurstTransfer, and DmaBufBlkSize as ExternalBurstSize or AxiBurstSize. This will allow us to:

- Default to the smallest block size by default for any external/axi accesses. The burst size shouldn't really matter in terms of compatibility, only in terms of performance.
- Allows us to provide device-specific configurability using the same function, and device-specific types. Devices that don't support PSRAM, we would only use the BurstTransfer enum as written here, or a variant of this for the C5 and C61.
- Users will be required to completely specify what they expect from a buffer, if they touch the burst mode config.
- The EDMA/AXI dma burst length will be changeable. While the S2 has a note (`Note that all EDMA should have the same block size.`) it's unclear what this refers to, since all EDMA peripherals have separate configuration fields.